[#220] Add deployment guide for Heroku (#651)
slackapi/bolt-js: docs/_deploying/heroku.md

---
title: Deploying to Heroku
order: 0
slug: heroku
lang: en
layout: tutorial
permalink: /deploying/heroku
redirect_from:
  - /deploying
---
# Deploying to Heroku

<div class="section-content">
This guide will walk you through preparing and deploying a Slack app using [Bolt for JavaScript](/bolt-js) and the [Heroku platform](https://heroku.com/). Along the way, we’ll download a Bolt Slack app, prepare it for Heroku, and deploy it.
</div>

When you’re finished, you’ll have this ⚡️[Getting Started with Heroku app](https://github.com/mwbrooks/bolt-js-getting-started-with-heroku-app) to run, modify, and make your own.

---

### Get a Bolt Slack app

> We'll continue from [Getting Started with Bolt for JavaScript][getting-started-guide] to learn how to prepare and deploy a Slack app to Heroku.

Reviewer: ```suggestion
If you haven't already built your own Bolt app, you can use our [getting started guide][getting-started-guide] or clone the template app below.
```

Author: Thank you! Fixed in b340032 because GitHub will not allow me to apply your suggestion (outdated file).

> Download the Getting Started app and change into its directory:

Reviewer: If you use the suggestion above, I'd remove this line and separate out the `clone` command and the `cd` command (I think most people will already have an existing app).

The `clone` command can go after the first sentence and then under that, you could add a new sentence like "After you have a Bolt app, navigate to its directory:"

Author: Fixed in 3c9ab8f. Thanks for this suggestion. I think it flows much better than my original attempt and it addressed my concern that developers using an existing app may forget to change into the app's directory.

```shell

> git clone https://github.com/mwbrooks/bolt-js-getting-started-app.git

Reviewer 2: How about having this repo in `github.com/slackapi`?

Author: I've now transferred my repository to `github.com/slackapi` and updated the references in commit 0cad3bb 🥳

cd bolt-js-getting-started-app/
```

> > 💡 You also can follow along with an existing Bolt app and we'll mention any extra steps that are needed.

Reviewer: I think you can remove this

Author: Agreed. Now that we have separated the two commands, it's more obvious that it's okay to use an existing app.

Fixed in 884c97c

> Now that you have the app, let's prepare it for Heroku.

Reviewer: ```suggestion
Now that you have an app, let's prepare it for Heroku.
```

Author: Fixed in 7ee5c1d

---

### Prepare the app for Heroku

> Heroku is a flexible platform that uses a few conventions to understand how to host your app. In this section, we'll update your app to use those conventions.

Author: Updated in 158887d. Originally, I was trying to use the "conventions over configuration" terminology, but I think your suggestion is more reader-friendly. Thanks!

**1. Use a Git repository**

> Before you can deploy your app to Heroku, you need to [install Git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git), initialize a local Git repository, and commit your code to it.

Reviewer: I think we can remove most of this section. I'd do something like:

Before you can deploy your app to Heroku, you'll need a git repository. If you aren't already using Git, you'll need to [install Git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git) and follow the instructions to [create a Git repository](https://git-scm.com/book/en/v2/Git-Basics-Getting-a-Git-Repository).

You can keep the > Skip this step... part, but I'd remove everything below that.

Author: Good suggestion! It shortens up this section for the majority of developers. I've applied this in 7ee5c1d

> 💡 Skip this step if you used `git clone` in the previous section because you already have a Git repository.

> To initialize a local Git repository and run the command:

Author: We ended up removing the Git commands after @shaydewael suggested a way to tighten up the Git initialization section.

```shell
git init
```

Now add your code and commit it to the repository:

```shell
git add .
git commit -m "Initial commit"
```

**2. Add a Procfile**

Every Heroku app uses a special file called `Procfile` that tells Heroku how to start your app. A Bolt Slack app will be started as a web server with a public web address.

> Create a new file called `Procfile` (without any extension) in your app's root directory with the following content:

Reviewer: Perhaps "with the following content" -> "paste the following"

Author: Paste is such a good word. It's silly that I didn't think of it. Fixed in 7ee5c1d

```yaml
web: node app.js
```

Once you've saved the file, let's commit it to your Git repository:

```shell
git add Procfile
git commit -m "Add Procfile"
```

> **3. Extra steps for existing apps**

Reviewer: Instead of a new section, I'd use a 💡 callout like you did above.

Author: I really like this suggestion. It helps to tighten up this lengthy section. Fixed in 7ee5c1d

> Are you following this guide with an existing app that you created? If so, please review [preparing a codebase for Heroku](https://devcenter.heroku.com/articles/preparing-a-codebase-for-heroku-deployment#4-listen-on-the-correct-port) to listen on the correct port and think about data storage provisioning.

Reviewer: ```suggestion
Are you following this guide with an existing Bolt app? If so, please review the guide on [preparing a codebase for Heroku](https://devcenter.heroku.com/articles/preparing-a-codebase-for-heroku-deployment#4-listen-on-the-correct-port).
```

Author: Thanks for making this suggestion. I was struggling to word this well. It's silly, but while writing the guide the term "Bolt app" didn't come to mind as a way to differentiate talking about your "Bolt app" and "Heroku app." Fixed in 7ee5c1d

---

### Setup the Heroku tools

Now we can set up the Heroku tools on your local machine. These tools will help you manage, deploy, and debug your app on Heroku's platform.

**1. Install the Heroku CLI**

The Heroku tools are available as a Command Line Interface (CLI). Go ahead and [install the Heroku CLI for macOS, Windows, or Linux](https://devcenter.heroku.com/articles/getting-started-with-nodejs#set-up). On macOS, you can run the command:

```shell
brew install heroku/brew/heroku
```

Once the install is complete, we can test the Heroku CLI by displaying all of the wonderful commands available to you:

```shell
heroku help
```

> > 💡 If the `heroku` command is not found, you can refresh your path by opening a new terminal session/tab.

Author: Nice! Fixed in 158887d

**2. Log into the Heroku CLI**

The Heroku CLI connects your local machine with your Heroku account. [Sign up for a free Heroku account](https://heroku.com) and then log into the Heroku CLI with the following command:

```shell
heroku login
```
> 💡 If you're behind a firewall, you may need to [set the proxy environment variables](https://devcenter.heroku.com/articles/using-the-cli#using-an-http-proxy) for the Heroku CLI.

**3. Confirm you're logged into the Heroku CLI**

> Let's check that you're logged in by displaying the account currently connected to your Heroku CLI:

Author: Good suggestion, thanks! Updated in 158887d

```shell
heroku auth:whoami
```

> You're all setup with the Heroku tools! Now let's move onto the exciting step of creating an app on Heroku.

Reviewer: ```suggestion
You should now be all setup with the Heroku tools! Let's move onto the exciting step of creating an app on Heroku.
```

Author: Applied @shaydewael suggestion in 7ee5c1d.

@misscoded This is a good point. At this point in the guide, we've only setup the Heroku tools on the local machine. In the next step, we will deploy the app. In that section, we do walk the developer through testing the app inside Slack.

---

### Create an app on Heroku

It’s time to [create a Heroku app](https://devcenter.heroku.com/articles/creating-apps) using the tools that you just installed. When you create an app, you can choose a unique name or have it randomly generated.

> > 💡 You can [rename a Heroku app](https://devcenter.heroku.com/articles/renaming-apps) at any time, but you may need to update your Git remote and app URL.

Reviewer: I think you can delete this

Author: @shaydewael if possible, I'd like to keep this section. In the Getting Started guide, I really appreciated the note that we can rename our Slack app after creating it because I was immediately struck with _naming paralysis_. I felt this call out would do the same to ease developers fears about naming their Heroku app.

If you strongly feel we can remove this section, just let me know and I'll delete it!

**1. Create an app on Heroku**

Create an app on Heroku with a unique name:

```shell
heroku create my-unique-bolt-app-name
```

or, have some fun with a random name:

```shell
heroku create
# Creating sharp-rain-871... done, stack is heroku-18
# https://sharp-rain-871.herokuapp.com/ | https://git.heroku.com/sharp-rain-871.git
```

After your app is created, there are a few things worth your attention:

- App name is `sharp-rain-871`
- Web address is `https://sharp-rain-871.herokuapp.com/`
- Empty Git remote is `https://git.heroku.com/sharp-rain-871.git`

**2. Confirm Heroku Git remote**

The Heroku CLI automatically adds a Git remote called `heroku` to your local repository. You can list your Git remotes to confirm `heroku` exists:

```shell
git remote -v
# heroku	https://git.heroku.com/sharp-rain-871.git (fetch)
# heroku	https://git.heroku.com/sharp-rain-871.git (push)
```

**3. Set environment variables on Heroku**

> In the Getting Started guide, you [exported your signing secret and token](http://localhost:4000/bolt-js/tutorial/getting-started#tokens-and-installing-apps) as environment variables. Now we need to set the same environment variables in your Heroku app:

Reviewer: ```suggestion
In the Getting Started guide, you [exported your signing secret and token](https://slack.dev/bolt-js/tutorial/getting-started#tokens-and-installing-apps) as environment variables. Now we need to set the same environment variables in your Heroku app:
```

Reviewer: This assumes people have followed the getting started guide. It may be better to say something like "Now you'll need to add your Slack app credentials to your Heroku app. If you don't know where to find your crendentials, read the getting started guide" (with the link you had). Or something that reframes it as the getting started guide coming after what the developer needs.

Author: Thanks for catching the `localhost` 🤦🏻 and I've updated the section with your rewording in commit 7ee5c1d. I like how you've chose to frame it - to focus on the next task rather than drawing parallels to the Getting Started guide.

```shell
heroku config:set SLACK_SIGNING_SECRET=<your-signing-secret>

> heroku config:set SLACK_BOT_TOKEN=xoxb-<your-bot-token>

Author: 100% agree with you and the same concern has crossed my mind. I chose to use the above syntax because it's what we use the Getting Started guides for JavaScript and Python. So, I thought I'd stay consistent.

However, I'll make a note to discuss how we can represent the Bot Token and then we can update all guides at once.

```

Now that we have prepared your local app and created a Heroku app, the next step is to deploy it!

---

### Deploy the app

To deploy the app, we're going to push your local code to Heroku, update your Slack app's settings, and say "hello" to your Heroku app. ✨

**1. Deploy the app to Heroku**

When deploying an app to Heroku, you'll typically use the `git push` command. This will push your code from your local repository to your `heroku` remote repository.

You can now deploy your app with the command:

```shell
git push heroku main
```

> > 💡 Heroku deploys code that's pushed to the [master or main branches](https://devcenter.heroku.com/articles/git-branches). Pushing to other branches does nothing.

Author: Thanks! Updated in 158887d

Now that your code is deployed, we need to start a web server instance using the Heroku CLI:

```shell
heroku ps:scale web=1
```

**2. Update your Slack app's setting**

> In the [Getting Started guide][getting-started-guide], we used ngrok as your **Request URL** for actions and events. Now we need to use your Heroku web address.

Reviewer: ```suggestion
Now we need to use your Heroku web address as your **Request URL**, which is where Slack will send events and actions.
```

Author: Much better! Fixed in 7ee5c1d

Get your Heroku web address with the following command:

```shell
heroku info
# ...
# Web URL: https://sharp-rain-871.herokuapp.com/
```

In our example, the web address is `https://sharp-rain-871.herokuapp.com/`.

> Now browse to the [Slack App page](https://api.slack.com/apps) and select your app name. Next, we'll update your **Request URL** in two locations to be your web address.

Author: Really great suggestion, so thank you. While writing, I felt like I was over-using "Now" as well. Updated in 158887d

> 💡 Your **Request URL** ends with `/slack/events`, such as `https://sharp-rain-871.herokuapp.com/slack/events`.

First, select **Interativity & Shortcuts** from the side and update the **Request URL**:

![Interactivity & Shortcuts page](../assets/interactivity-and-shortcuts-page.png "Interactivity & Shortcuts page")

Second, select **Event Subscriptions** from the side and update the **Request URL**:

![Event Subscriptions page](../assets/event-subscriptions-page.png "Event Subscriptions page")

> 💡 Free Heroku apps sleep when inactive. 💤 If your verification fails, please try it again immediately.

**3. Test your Slack app**

Your app is now deployed and Slack is updated, so let's try it out!

> Open a Slack channel that your app has joined and say "hello" (lower-case). Just like in the [Getting Started guide](/bolt-js/tutorial/getting-started#sending-and-responding-to-actions), your app should respond back.

Author: This also crossed my mind. Originally, I didn't elaborate because the Getting Started guide didn't. However, I've now added an extra sentence suggestion that the developer double-checks their **Request URL**. Updated in 158887d

As you continue building your Slack app, you'll need to deploy updates. A common flow is to make a change, commit it, and then push it to Heroku.

<!-- TODO - Update GitHub link -->
Let's get a feel for this by updating your app to respond to a "goodbye" message. Add the following code to `app.js` ([source code on GitHub](https://github.com/mwbrooks/bolt-js-getting-started-app/blob/master/app.js)):

```javascript
// Listens to incoming messages that contain "goodbye"
app.message('goodbye', async ({ message, say }) => {
  // say() sends a message to the channel where the event was triggered
  await say(`See ya later, <@${message.user}> :wave:`);
});
```

Commit the changes to your local Git respository:

```shell
git commit -am "Say 'goodbye' to a person"
```

Deploy the update by pushing to your `heroku` remote:

```shell
git push heroku main
```

When the deploy is complete, you can open a Slack channel that your app has joined and say "goodbye" (lower-case). You should see a friendly farewell from your Slack app.

---

### Next steps

You just deployed your first ⚡️[Bolt for JavaScript app to Heroku](https://github.com/mwbrooks/bolt-js-getting-started-with-heroku-app)! 🚀

> Now that you've deployed a basic app, you can start exploring how to customize and monitor it. Here are some ideas of what to explore next:

Reviewer: I like that you've included Heroku call to actions. Could we also add CTA to read through Bolt documentation?

Author: Good call! I don't know why I overlooked that. I've updated the Next Steps to have 4 bullet points that include the Basic and Advanced concepts along with important Heroku next steps. Fixed in 7ee5c1d

---
title: Deploying to Heroku
order: 0
slug: heroku
lang: en
layout: tutorial

> permalink: /deploying/heroku

Author: The URL will be `/deploying/heroku` unless there is a better suggestion.

Reviewer 2: I prefer `/deployments/heroku` or `/deployments-heroku`

Author: Awesome, this aligns well with some feedback from @shaydewael to use "Deployments" for the header section. I've now updated the URL to be `/deployments/heroku`

Fixed in f628553

redirect_from:
  - /deploying
---
# Deploying to Heroku

<div class="section-content">
This guide will walk you through preparing and deploying a Slack app using [Bolt for JavaScript](/bolt-js) and the [Heroku platform](https://heroku.com/). Along the way, we’ll download a Bolt Slack app, prepare it for Heroku, and deploy it.
</div>

> When you’re finished, you’ll have this ⚡️[Getting Started with Heroku app](https://github.com/mwbrooks/bolt-js-getting-started-with-heroku-app) to run, modify, and make your own.

Author: @shaydewael while writing this guide, I was curious about your preference on using contractions like _you're_ and _you'll._ Do you try to avoid contractions or use them liberally?

Reviewer: I'm fine with contractions! Sometimes I'll avoid them for clarity, but besides that I use them plenty.

Author: Great to know! This will be super helpful to me while I write the other deployment guides 🙇🏻

> When you’re finished, you’ll have this ⚡️[Getting Started with Heroku app](https://github.com/mwbrooks/bolt-js-getting-started-with-heroku-app) to run, modify, and make your own.

Author: I'll open a team discussion to talk about where the example apps should live - separate repository, subfolder in bolt, or elsewhere. Afterwards, I'll update the links in this guide.

Author: Added the _Getting Started with Heroku_ app to the `examples/` directory. Commit e34e905

---

### Get a Bolt Slack app

We'll continue from [Getting Started with Bolt for JavaScript][getting-started-guide] to learn how to prepare and deploy the Getting Started app to Heroku.

Download the Getting Started app and change into its directory:

```shell
git clone https://github.com/mwbrooks/bolt-js-getting-started-app.git
cd bolt-js-getting-started-app/

> ```

Author: Question: every command after this section assumes that you're inside the app directory from running `cd bolt-js-getting-started-app/`. However, some users may be reading along with their own app and may not run this command. That could cause issues later on in the guide.

An especially problematic issue would be initializing a git repository in a directory that's not the app directory.

Do you think it's worth adding a note that we assume every command from here on is run inside the app directory? This wasn't called out in the _Getting Started_ guide, which is why I didn't in this guide.

Reviewer 2: I think most developers don't have the issue but if you could have such a kind guide here, it'll be much more beginner-friendly.

Author: Based on @shaydewael's feedback, we separated out these two commands. This makes it easier to follow for developers who are using an existing Bolt app and clarifies the importance of changing into the app's directory.

Let's check that you're logged in by displaying the account currently connected to your Heroku CLI:

```shell
heroku auth:whoami
```

You're all setup with the Heroku tools! Now let's move onto the exciting step of creating an app on Heroku.

---

### Create an app on Heroku

It’s time to [create a Heroku app](https://devcenter.heroku.com/articles/creating-apps) using the tools that you just installed. When you create an app, you can choose a unique name or have it randomly generated.

> 💡 You can [rename a Heroku app](https://devcenter.heroku.com/articles/renaming-apps) at any time, but you may need to update your Git remote and app URL.

**1. Create an app on Heroku**

Create an app on Heroku with a unique name:

```shell
heroku create my-unique-bolt-app-name
```

or, have some fun with a random name:

```shell
heroku create
# Creating sharp-rain-871... done, stack is heroku-18
# https://sharp-rain-871.herokuapp.com/ | https://git.heroku.com/sharp-rain-871.git
# Git remote heroku added
```

After your app is created, there are a few things worth your attention:

- App name is `sharp-rain-871`
- Web address is `https://sharp-rain-871.herokuapp.com/`

> - Empty Git remote is `https://git.heroku.com/sharp-rain-871.git`

Author: I was on the fence whether this is helpful or not. Later on in the guide, we do reference this example values, which is why I wanted to call them out here.

Author: This is something that crossed my mind as well. Taking your advice, I've updated it to foreshadow using the info in upcoming sections. Updated in 158887d

Now that your code is deployed, we need to start a web server instance using the Heroku CLI:

```shell
heroku ps:scale web=1
```

**2. Update your Slack app's setting**

In the [Getting Started guide][getting-started-guide], we used ngrok as your **Request URL** for actions and events. Now we need to use your Heroku web address.

Get your Heroku web address with the following command:

```shell
heroku info
# ...
# Web URL: https://sharp-rain-871.herokuapp.com/
```

In our example, the web address is `https://sharp-rain-871.herokuapp.com/`.

Now browse to the [Slack App page](https://api.slack.com/apps) and select your app name. Next, we'll update your **Request URL** in two locations to be your web address.

> 💡 Your **Request URL** ends with `/slack/events`, such as `https://sharp-rain-871.herokuapp.com/slack/events`.

First, select **Interativity & Shortcuts** from the side and update the **Request URL**:

![Interactivity & Shortcuts page](../assets/interactivity-and-shortcuts-page.png "Interactivity & Shortcuts page")

Second, select **Event Subscriptions** from the side and update the **Request URL**:

> ![Event Subscriptions page](../assets/event-subscriptions-page.png "Event Subscriptions page")

Author: @shaydewael I added two images to help explain where to update the **Request URL** in the app management page. I know you're thinking about improving the images in the _Getting Started_ guide. I don't want these images to step on the toes of your work, so let me know if there's a better approach that you have in mind.

> 💡 Free Heroku apps sleep when inactive. 💤 If your verification fails, please try it again immediately.

**3. Test your Slack app**

Your app is now deployed and Slack is updated, so let's try it out!

Open a Slack channel that your app has joined and say "hello" (lower-case). Just like in the [Getting Started guide](/bolt-js/tutorial/getting-started#sending-and-responding-to-actions), your app should respond back.

---

### Deploy an update

As you continue building your Slack app, you'll need to deploy updates. A common flow is to make a change, commit it, and then push it to Heroku.

<!-- TODO - Update GitHub link -->

> Let's get a feel for this by updating your app to respond to a "goodbye" message. Add the following code to `app.js` ([source code on GitHub](https://github.com/mwbrooks/bolt-js-getting-started-app/blob/master/app.js)):

Author: I'll open a team discussion to talk about where the example apps should live - separate repository, subfolder in bolt, or elsewhere. Afterwards, I'll update the links in this guide.

Author: The _Getting Started with Heroku_ example app is now bundled in the `examples/` directory. Commit e34e905

```javascript
// Listens to incoming messages that contain "goodbye"
app.message('goodbye', async ({ message, say }) => {
  // say() sends a message to the channel where the event was triggered
  await say(`See ya later, <@${message.user}> :wave:`);
});
```

Commit the changes to your local Git respository:

```shell
git commit -am "Say 'goodbye' to a person"
```

Deploy the update by pushing to your `heroku` remote:

```shell
git push heroku main
```

When the deploy is complete, you can open a Slack channel that your app has joined and say "goodbye" (lower-case). You should see a friendly farewell from your Slack app.

---

### Next steps

> You just deployed your first ⚡️[Bolt for JavaScript app to Heroku](https://github.com/mwbrooks/bolt-js-getting-started-with-heroku-app)! 🚀

Author: I'll open a team discussion to talk about where the example apps should live - separate repository, subfolder in bolt, or elsewhere. Afterwards, I'll update the links in this guide.

Reviewer 2: I think we can have it in a subfolder in bolt or as a separate repo in github.com/slackapi

Author: Added the _Getting Started with Heroku_ app to the `examples/` directory in commit e34e905
